core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+    println("Error sending segment performance metrics $err")
+}
+
+object Telemetry {
+    private var _enable: Boolean = true
+    private var _host: String = Constants.DEFAULT_API_HOST
+    private var _sampleRate: Double = 1.0
+    private var _flushTimer: Int = 30 * 1000 // 30s
+    private var _maxQueueSize: Int = 20
+    private var MAX_QUEUE_BYTES = 28000
+    private var _maxQueueBytes: Int = MAX_QUEUE_BYTES
+    private var _httpClient: HTTPClient = HTTPClient("", MetricsRequestFactory())
+    private var _sendWriteKeyOnError: Boolean = true
+    private var _sendErrorLogData: Boolean = false
+    private var _errorHandler: ((Throwable) -> Unit)? = null
+    private var _rateLimitTimer: Int = 0
+
+    var enable: Boolean
+        get() = _enable
+        set(value) {
+            _enable = value
+        }
+    var sendWriteKeyOnError: Boolean
+        get() = _sendWriteKeyOnError
+        set(value) {
+            _sendWriteKeyOnError = value
+        }
+
+    var sendErrorLogData: Boolean
+        get() = _sendErrorLogData
+        set(value) {
+            _sendErrorLogData = value
+        }
+    var errorHandler: ((Throwable) -> Unit)?
+        get() = _errorHandler
+        set(value) {
+            _errorHandler = value
+        }
+
+    var host: String
+        get() = _host
+        set(value) {
+            _host = value
+        }
+
+    var sampleRate: Double
+        get() = _sampleRate
+        set(value) {
+            _sampleRate = value
+        }
+
+    var flushTimer: Int
+        get() = _flushTimer
+        set(value) {
+            _flushTimer = value
+        }
+
+    var maxQueueSize: Int
+        get() = _maxQueueSize
+        set(value) {
+            _maxQueueSize = value
+        }
+
+    var maxQueueBytes: Int
+        get() = _maxQueueBytes
+        set(value) {
+            _maxQueueBytes = min(value, MAX_QUEUE_BYTES)
+        }
+
+    var httpClient: HTTPClient
+        get() = _httpClient
+        set(value) {
+            _httpClient = value
+        }
+
+    private val queue = mutableListOf<RemoteMetric>()
+    private var queueBytes = 0
+    private var queueSizeExceeded = false
+    private val seenErrors = mutableMapOf<String, Int>()
+    private var started = false
+
+    init {
+        CoroutineScope(Dispatchers.Default).launch {
+            while (isActive) {
+                if (!enable) return@launch
+                try {
+                    flush()
+                } catch (e: Throwable) {
+                    logError(e)
+                }
+                try {
+                    delay(flushTimer.toLong())
+                } catch (e: CancellationException) {
+                    // last flush before shutdown
+                    flush()
+                }
+
+            }
+        }
+    }
+
+    fun start() {
+        if (started) return
+        // Assume sampleRate is now set and everything in the queue hasn't had it applied
+        if (Math.random() > sampleRate) {
```

Comment:
`>=` no ?

Reply:
It's following the logic in analytics_next.  
Hrm.. if the sampleRate is 0.0 from settings, >= would let it send data (on a 4e15th chance)
But actually, I should add logic to just not start if sample rate is zero, either way.  👍 
... Also because I just realized I'm using sampleRate as the divisor for value.  Glad I looked!  More checks added.